### PR TITLE
Injection before and after as kwargs

### LINKED ISSUE #22: Specify what cell to inject `before` or `after` as kwargs
I think we need to specify what cell to inject before or after as kwargs. 

_Originally posted by @MSeal in https://github.com/_render_node/MDIzOlB1bGxSZXF1ZXN0UmV2aWV3VGhyZWFkMjQ5NzY5MTYwOnYy/pull_request_review_threads/discussion_

### PR DESCRIPTION
**Pull request by rohitsanj**
Added `before` and `after` functionality to `inject`, which injects before or after a specified cell.

Also added a property method called `cells` which makes it easier to access the cells, so instead of `tb.nb.cells`, it can be directly accessed with `tb.cells` ( where`tb` is an instance of `TestbookNotebookClient`)

resolves #22

### CODE REVIEW

File: testbook/client.py [hunk 1]
```diff
@@ -12,6 +12,13 @@
 
 
 class TestbookNotebookClient(NotebookClient):
+    def __init__(self, nb, km=None, **kw):
+        super().__init__(nb, km=km, **kw)
+
+    @property
+    def cells(self):
```
Comment (MSeal) [inline testbook/client.py:19]: :+1:

File: testbook/client.py [hunk 2]
```diff
@@ -123,11 +130,20 @@ def inject(self, code, args=None, prerun=None):
         else:
             raise TypeError('can only inject function or code block as str')
 
+        inject_idx = len(self.cells)
+
+        if kwargs.get("after") is not None and kwargs.get("before") is not None:
```
Comment (MSeal) [inline testbook/client.py:135]: Why not just have `before=None, after=None` as named arguments?

Comment (rohitsanj) [reply testbook/client.py:135]: Sure, makes sense.

**Comment by rohitsanj** [comment]
> So is the benefit to have the sources relatively assigned for export purposes? Do we need to make run-up-to/through-cell a more prevalent pattern to help ?

I think an `execute_upto_cell` would certainly help. Will add that in a future PR. 
Along with an `execute_notebook` that executes the entire notebook. 

I think the benefit of this is to properly inject cells in order, but since we're executing as soon as we inject - makes me think if we should rather just inject so that we'd be able to run all cells linearly up to (and including or excluding) the injected cell. I think this would make it more intuitive.

**Comment by MSeal** [comment]
Yeah I think maybe make the inject take a flagged argument to run immediately, but perhaps default to lazy injection without execution would be more useful.

**Comment by rohitsanj** [comment]
Changes:
- added a boolean argument called `run` which controls whether the injected cell should be immediately executed or not. 

- removed `prerun` argument from inject for two reasons
   - the `prerun` arguments is also present in `testbook.testbook` decorator and might confuse users.
   - it might be confusing to have arguments called `run` and `prerun` to the same function.

File: testbook/client.py [hunk 3]
```diff
@@ -123,24 +135,30 @@ def inject(self, code, args=None, prerun=None):
         else:
             raise TypeError('can only inject function or code block as str')
 
-        if prerun is not None:
-            self.execute_cell(prerun)
+        inject_idx = len(self.cells)
+
+        if after is not None and before is not None:
+            raise TypeError("pass either before or after as kwargs")
```
Comment (MSeal) [inline testbook/client.py:141]: Should be a ValueError here and not a TypeError

File: testbook/tests/test_inject.py [hunk 4]
```diff
@@ -64,37 +64,16 @@ def test_inject_raises_exception():
...
-            "Hello there\nBye",
-        ),
-    ],
-)
-def test_inject_with_prerun(prerun, code_block, expected_text):
-    with testbook('testbook/tests/resources/inject.ipynb') as notebook:
-        assert notebook.inject(code_block, prerun=prerun).output_text == expected_text
+def test_inject_before_after():
+    with testbook('testbook/tests/resources/inject.ipynb', prerun=['hello', 'bye']) as notebook:
+        notebook.inject("say_hello()", run=True, after="hello")
```
Comment (MSeal) [inline testbook/tests/test_inject.py:72]: I'd leave run=False since you're testing the injection and not the execution?

**Pull request merged**